include/deal.II/non_matching/fe_values.h

```diff
@@ -0,0 +1,419 @@
+     * optional will not contain a value.
+     */
+    const std_cxx17::optional<dealii::FEValues<dim>> &
+    get_inside_fe_values() const;
+
+    /**
+     * Return an dealii::FEValues object reinitialized with a quadrature for the
+     * outside region of the cell: $\{x \in K : \psi(x) > 0 \}$.
+     *
+     * @note If the quadrature rule over the region is empty the returned
+     * optional will not contain a value.
+     */
+    const std_cxx17::optional<dealii::FEValues<dim>> &
+    get_outside_fe_values() const;
+
+    /**
+     * Return an dealii::FEValues object reinitialized with a quadrature for the
+     * surface region of the cell: $\{x \in K : \psi(x) = 0 \}$.
+     *
+     * @note If the quadrature rule over the region is empty the returned
+     * optional will not contain a value.
+     */
+    const std_cxx17::optional<FEImmersedSurfaceValues<dim>> &
+    get_surface_fe_values() const;
+
+  private:
+    /**
+     * Do work common to the constructors. The incoming QCollection should be
+     * quadratures integrating over $[0, 1]^{dim}$. These will be used on the
+     * non-intersected cells.
+     */
+    void
+    init(const hp::QCollection<dim> &q_collection);
+
+    /**
+     * A pointer to the collection of mappings to be used.
+     */
+    const SmartPointer<const hp::MappingCollection<dim>> mapping_collection;
+
+    /**
+     * A pointer to the collection of finite elements to be used.
+     */
+    const SmartPointer<const hp::FECollection<dim>> fe_collection;
+
+    /**
+     * Collection of 1-dimensional quadrature rules that are used by
+     * QuadratureGenerator as base for generating the immersed quadrature rules.
+     */
+    const hp::QCollection<1> q_collection_1D;
+
+    /**
+     * Location of the last cell that reinit was called with.
+     */
+    LocationToLevelSet current_cell_location;
+
+    /**
+     * Active fe index of the last cell that reinit was called with.
+     */
+    unsigned int active_fe_index;
+
+    /**
+     * The update flags passed to the constructor.
+     */
+    const RegionUpdateFlags region_update_flags;
+
+    /**
+     * Pointer to the MeshClassifier passed to the constructor.
+     */
+    const SmartPointer<const MeshClassifier<dim>> mesh_classifier;
+
+    /**
+     * For each element in the FECollection passed to the constructor,
+     * contains an dealii::FEValues object created with a quadrature rule over
+     * the full reference cell: $[0, 1]^dim$ and UpdateFlags for the inside
+     * region. Thus, these optionals should always contain a value.
+     *
+     * When LocationToLevelSet of the cell is INSIDE (and we don't need
+     * to generate an immersed quadrature), we return the dealii::FEValues
+     * object in this container corresponding to the cells active_fe_index.
+     *
+     * This container is a std::deque, because most functions in std::vector
+     * need the contained class to have a copy-constructor.
```

Comment:
```suggestion
     * This container is a std::deque, which is compatible with the `FEValues` class
     * that does not have a copy-constructor.
```
(Use the same in the places below.)